Feature/scrum 57 back appointment confirmation

LeonelCeballos18 opened the pull request:
## What kind of PR is this?
- [x] Feature

## Description
-add formatter message
-add `update status` and `get last appointment` in appointment service
-update status in `whatsapp controller`

## Ticket id
https://barbershop-app.atlassian.net/jira/software/projects/SCRUM/boards/1?selectedIssue=SCRUM-57

## Useful resources
![image](https://github.com/user-attachments/assets/6a7dbe2f-f500-4148-8fa6-b5a27b62fe85)
![image](https://github.com/user-attachments/assets/c66d1f82-d701-451f-bd3c-3b601457e30f)

LeonelCeballos18 committed:
to work at home :)

LeonelCeballos18 committed:
feature: Appointment confirmation

-add formatter message
-add `update status` and `get last appointment` in appointment service
-update status in `whatsapp controller`

LeonelCeballos18 committed:
fix: format month

-fix format month in `messageFormatter`

cmoreno357 commented on src/utils/messageFormatter.ts:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  if (type === "appointment_canceled") {
+    params["url"] = "https://mindticas2025-front-git-development-michelada-interns.vercel.app/"; //URL development
```
Please use this as an environment variable.

LeonelCeballos18 replied:
got it :)

LeonelCeballos18 committed:
chore: change error messages to english

LeonelCeballos18 committed:
chore: Solving comments

-add enviroment variable for url development

cmoreno357 commented on package.json:
```diff
@@ -32,6 +32,7 @@
       "class-transformer": "^0.5.1",
       "class-validator": "^0.14.1",
       "dotenv": "^16.4.7",
+      "ngrok": "^5.0.0-beta.2",
```
@cmagallon1 I reckon this is a devDependency, isn't it? What do you think?

cesarj45 replied:
yarn remove ngrok

cesarj45 replied:
yarn add -D ngrok

LeonelCeballos18 replied:
solved 😎

cesarj45 committed:
feature: Clients | Remainder of an appointment

- Add `schedule` dependency to implement cron jobs for setting reminders.
- Add `scheduleReminderMessage` method to start the cron job and allow users to cancel appointments.
- Implement `OnModuleInit` interface to restore all reminders when the app is restarted.
- Improve error handling for missing environment variables.
- Refactor `sendInteractiveMessage` buttons to make them modular.
- Fix the `WhatsAppController` to send an ACK.

LeonelCeballos18 committed:
chore: Solving comments

-moving ngrok to DevDependencies

cesarj45 reviewed (approved):
LGTM

mtopete1996 commented on src/controllers/whatsapp.controller.ts:
```diff
@@ -11,23 +12,41 @@ export default class TreatmentController {
-          messages['appointment_canceled'],
-        );
+  @HttpCode(200)
+  async handleWebhook(@Body() body: any) {
```
@cmoreno357  @cmagallon1  idk how JS works, but I'd put all this logic into another class/module/service. Functions with > 8-10 lines trigger my SOLID instincts

dsalazar357 replied:
It's the same thing... Should b split in different concerns

LeonelCeballos18 replied:
Cesar is gonna solve this in his PR

LeonelCeballos18 replied:
I already solved this

mtopete1996 replied:
Good job reducing nested ifs, but you're missing the "put all this logic into another class/module/service" part

LeonelCeballos18 replied:
solved

mtopete1996 commented on src/services/appointment.service.ts:
```diff
@@ -86,11 +87,12 @@ export default class AppointmentService {
 
     try {
```
Same with this create method, but I wouldn't do it in this PR

LeonelCeballos18 committed:
chore: Solving commets

-optimaze `create` method in `appointment service`

LeonelCeballos18 committed:
chore: Solving comments

-optimize `whatsapp.controller`
-fix `getLastAppointmetByPhone` to order the customer appointments by the last id

cesarj45 reviewed (approved):
LGTM

mtopete1996 reviewed (changes requested):
Missing changes

LeonelCeballos18 committed:
chore: Solving comments

-move `handleWebhook` to `whatsapp.service`

LeonelCeballos18 committed:
feature: appointment cancellation

-add appointment cancellation in `whatsapp.service`

cmoreno357 committed:
Merge pull request #28 from cmoreno357/feature/SCRUM-14-appointment-cancellation

feature: appointment cancellation

cesarj45 committed:
Co-authored-by: Jorge Torres <Jorge-Antonio-Torres-Lozoya@users.noreply.github.com>

cmoreno357 committed:
Merge pull request #27 from cmoreno357/feature/SCRUM-13-clients-remainder-of-an-appointment

feature: Clients | Remainder of an appointment

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -31,74 +41,145 @@ export default class AppointmentService {
       );
-      return saved;
+
+      try {
```
💭  why using a `try` inside a `try`? 🤔

I think the first try block can catch these errors too, you just need to create a catch block for each error:

```js
catch (Error1) {
  ...
} catch (Error2) {
  ...
}
```

cmoreno357 replied:
![donald-duck-angry](https://github.com/user-attachments/assets/91dda5b1-3921-4941-a689-91f131c1cf47)

cmoreno357 replied:
> 💭 why using a `try` inside a `try`? 🤔
> 
> I think the first try block can catch these errors too, you just need to create a catch block for each error:
> 
> ```js
> catch (Error1) {
>   ...
> } catch (Error2) {
>   ...
> }
> ```

![cigarrillos-para](https://github.com/user-attachments/assets/abc4d3ea-2492-428b-b8cf-76d26dd1f359)

cesarj45 replied:
We use a try block inside another because, in the second one, we need to handle the WhatsApp functionality. However, if it fails for some reason, such as a lack of payment to WHAPI, we still need to attempt to save the appointment without returning a 500 Internal Server Error. A 500 error should only be returned if the repository operation fails.

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -31,74 +41,145 @@ export default class AppointmentService {
+        );
+        this.scheduleReminderMessage(savedAppointment);
+      } catch (error) {
+        this.logger.error(`Error al conectar con whapi: ${error.message}`);
```
🎒 This error is logged in spanish but the one in line 85 is in English, should this one be in english too? 🤔

cmoreno357 replied:
> 🎒 This error is logged in spanish but the one in line 85 is in English, should this one be in inglish too? 🤔

Respaldo:
![Screenshot_2025-03-19-02-14-06-45_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/user-attachments/assets/14389c2d-db92-4625-a988-0c9465061172)

cmagallon1 replied:
upsi

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -31,74 +41,145 @@ export default class AppointmentService {
+        this.logger.log(
+          `Tiempo excedido para el recordatorio de la cita: ${appointment.id}.`,
+        );
+        return;
```
💭 if you're being explicit in this line about returning [null](https://github.com/cmoreno357/mindticas2025-back/pull/26/files#diff-ffade9a36b9a2d53d1cecc3eebc25f8abf4f715f8dc259e887fe3a30d23ea446R97)

shouldn't be more explicit here to follow the same standard? 🤔 

unless we want to remove the null in line 97

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -31,74 +41,145 @@ export default class AppointmentService {
+        return;
+      }
+
+      const job = new CronJob(reminderTime, async () => {
```
💅🏽 No need to address it in this PR, but this logic can be moved to another service class for appointment scheduling. There you can include all the logic related to the cronjob. Remember separating responsibilities and not being limited to a single class, you can create as many classes/methods as you need

cmoreno357 replied:
> 💅🏽 No need to address it in this PR, but this logic can be moved to another service class for appointment scheduling. There you can include all the logic related to the cronjob. Remember separating responsibilities and not being limited to a single class, you can create as many classes/methods as you need

I ran out of gif's but yes, what he said

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -113,31 +194,90 @@ export default class AppointmentService {
+      where: { scheduled_start: scheduledStart },
+    });
+
+    if (existingAppointment) {
```
💅🏽 

```suggestion
    if (!existingAppointment) { return; }
    
     throw new BadRequestException(
       'An appointment is already scheduled at this time.',
     );
```

cmagallon1 commented on src/services/appointment.service.ts:
```diff
@@ -113,31 +194,90 @@ export default class AppointmentService {
+      where: { phone: createDto.phone },
+    });
+
+    if (!customer) {
```
```suggestion
    if (customer) { return customer; }
    
    const customerDto = new CustomerRegisterDto();
    customerDto.name = createDto.name;
    customerDto.phone = createDto.phone;
    customer = await this.customerService.createCustomer(customerDto);
```

cmagallon1 commented on src/services/whatsapp.service.ts:
```diff
@@ -88,6 +101,67 @@ export default class WhatsAppService {
+    if (!messages) return;
+    const { id, type, from, reply } = messages[0];
+    this.sentACK(id);
+    if (type !== 'reply' || !reply?.buttons_reply?.id) return;
```
💅🏽 

```suggestion
    if (type !== 'reply' || !reply?.buttons_reply?.id) return;
    
```

cmagallon1 commented on src/services/whatsapp.service.ts:
```diff
@@ -88,6 +101,67 @@ export default class WhatsAppService {
+
+  async handleWebhook(body: any): Promise<any> {
+    const { messages } = body;
+    if (!messages) return;
```
💅🏽 

```suggestion
    if (!messages) return;
    
```

cmagallon1 commented on src/services/whatsapp.service.ts:
```diff
@@ -88,6 +101,67 @@ export default class WhatsAppService {
+    if (buttonId === 'ButtonsV3:2') {
+      await this.appointmentService.updateStatus(
+        lastAppointment.id,
+        Status.CONFIRMED,
```
💭 shouldn't this status be cancelled? 👀

LeonelCeballos18 replied:
you're right

cesarj45 committed:
Resolving comments

mtopete1996 reviewed (approved):
Good enough to me, solve pending comments 👍🏼
